File: src/main/kotlin/alg03sorting/S4aMergeSort.kt
package alg03sorting

import alg02deques.ArrayDeque
import alg02deques.Deque
import alg03ordered.Ordered

/**
 * Divide and conquer (split-sort-merge) algorithm.
 * The execution time is as fast as O(N*log2(N)).
 * Traditional recursion implementation.
 */
object S4aMergeSort : S4AbstractMergeSort() {
    private fun <T> sort(ordered: Ordered<T>, auxDeque: Deque<T>, comparator: Comparator<T>, lo: Int, hi: Int) {
        // If the indexes overlap the sorting is done.
        if (hi <= lo) {
            return
        }
        // Divide and conquer in action: the array is sliced in two parts...
        val mid = lo + (hi - lo) / 2
        // ...and each part is sorted independently...
        sort(ordered, auxDeque, comparator, lo, mid)
        sort(ordered, auxDeque, comparator, mid + 1, hi)
        // ... and two parts are merged back together.
        merge(ordered, auxDeque, comparator, lo, mid, hi)
    }

    override fun <T> sort(ordered: Ordered<T>, comparator: Comparator<T>) {
        val auxDeque = ArrayDeque<T>(ordered.size)
        for (item in ordered) {
            auxDeque.addLast(item)
        }
        this.sort(ordered, auxDeque, comparator, lo = 0, hi = ordered.size - 1)
    }
}

File: alg031sorting/s4a_merge_sort.go
package alg031sorting

import "github.com/deemson/algorithms/alg020indexed"

// MergeSort is a divide and conquer (split-sort-merge) algorithm.
// The execution time is as fast as O(N*log2(N)).
// Traditional implementation via recursion.
func MergeSort[T any](indexed alg020indexed.Indexed[T], less LessFunc[T]) {
	aux := alg020indexed.SliceAdapter[T]{
		Slice: make([]T, indexed.Size()),
	}
	mergeSort(indexed, aux, less, 0, indexed.Size()-1)
}

func mergeSort[T any](indexed, aux alg020indexed.Indexed[T], less LessFunc[T], lo, hi int) {
	// If the thresholds overlap the sorting is done.
	if lo >= hi {
		return
	}
	// Divide and conquer in action: the array is sliced in two parts...
	mid := lo + (hi-lo)/2
	// ...and each part is sorted independently...
	mergeSort(indexed, aux, less, lo, mid)
	mergeSort(indexed, aux, less, mid+1, hi)
	// ... and the two parts are merged back together.
	merge(indexed, aux, less, lo, mid, hi)
}
